### PR TITLE
add draft post for external-domain migration

### LINKED ISSUE #25: Document migration for users
We need to document for cdn service users what they need to do, when, and why.

### PR DESCRIPTION
**Pull request by bengerman13**
## Changes proposed in this pull request:
- add announcement about external-domain migration
- fixes cloud-gov/external-domain-broker-migrator#25
We should _not_ post this until we're ready for folks to start updating DNS.

:sunglasses:[PREVIEW URL](https://cg-88d42ca6-59d7-47e0-9500-4dd9251360b9.app.cloud.gov/preview/cloud-gov/cg-site/draft-migration-announcement)

## Security Considerations
None

### CODE REVIEW

**Comment by mheadd** [comment]
May want to break up the `What you need to do` section into individual paragraphs to make it easier to read.

Also, wondering if we want to add a standard closing paragraph like:

```
Customers that experience issues, or that have questions about this change, can send a request to support@cloud.gov.
```

or something similar?

**Comment by bengerman13** [comment]
Both of those suggestions sound good, Mark. I'll work them in

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+The new external-domain service uses a different method of validation with Let's Encrypt, which relies on specific DNS records being created on
+your part. To begin the migration, we need you to create 2 DNS CNAME records for each domain on your custom-domain or cdn-route instances.
+
+The first CNAME should created for the record `_acme-challenge.$DOMAIN` and should have the value
+`_acme-challenge.$DOMAIN.external-domains-production.cloud.gov`.  In other words, if you have a custom-domain service created for
+`directorate.agency.gov`, you'd want to create a CNAME record `_acme-challenge.directorate.agency.gov` with value
+`_acme-challenge.directorate.agency.gov.external-domains-production.cloud.gov`.
+
+The second CNAME should be for `$DOMAIN` and should point to `$DOMAIN.external-domains-production.cloud.gov`.  Using the `directorate.agency.gov`
+example above, this would be `directorate.agency.cloud.gov.external-domains-production.cloud.gov`.  Note that the second change here changes how
```
Comment (davemcorwin) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Can we make this a bit more explicit/simple that changing this CNAME will NOT cause downtime for users? This gets a bit technical for a lot of folks.

Comment (bengerman13) [reply _posts/2021-06-10-external-domain-migration-announcement.md]: I'm a little concerned about a blanket "this is safe" statement - updating your site's is something that should be approached with caution. The target audience here is the person/people making the DNS change, so I think the technical language is appropriate (but I'd happily accept suggestions on making it clearer). 
Would a "reach out with any questions" here make you feel better about this?

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+---
+layout: post
+title: "Migrate to the external-domain service"
+date: June 10, 2021
+excerpt: Customers need to take action to migrate from the cdn-service and custom-domain service to the external-domain service
+---
+
+## Who is this for?
+
+This post is for all users of the cdn-route or custom-domain services.
```
Comment (ccostino) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Should we include a specific qualifier here so people know these are deprecated services and to remind them?  e.g.:

```suggestion
This post is for all users of the deprecated cdn-route or custom-domain services.
```

File: _posts/2021-06-10-external-domain-migration-announcement.md (hunk 1 above)
Comment (bengerman13) [reply _posts/2021-06-10-external-domain-migration-announcement.md]: @davemcorwin I did a little wordsmithing here to try to make it more readable. LMK if you think it still needs work!

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+The custom-domain and cdn-route services leverage Let's Encrypt to provision certificates on our users' behalf.
+Several months ago, Let's Encrypt announced that they're deprecating their v1 API. To work with their new API, we've written a replacement
+to the cdn-route and custom-domain services, the external-domain service. We now need to migrate your service instances
+to the new service to ensure their certificates can continue renewing without issue. To accomplish this, we've written internal tooling to
+migrate instances automatically without service interruption, but it does require some action on your part to intiate the migration.
+
+## What you need to do
+
+The new external-domain service uses a different method of validation with Let's Encrypt. The new method relies on specific DNS records being created on
+your part. To begin the migration, we need you to create 2 DNS CNAME records for each domain on your custom-domain or cdn-route instances.
```
Comment (ccostino) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Would this read easier?

```suggestion
The new external-domain service uses a different method of validation with Let's Encrypt. The new method relies on specific DNS records being created for your domain(s). To begin the migration, you need to create 2 DNS CNAME records for each domain on your custom-domain or cdn-route instances.
```

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+The first CNAME should created for the record allows us to provision SSL certificates on your behalf.
```
Comment (ccostino) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Missing word:

```suggestion
The first CNAME should be created for the record allows us to provision SSL certificates on your behalf.
```

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+This record is for the name `_acme-challenge.$DOMAIN` and should have the value `_acme-challenge.$DOMAIN.external-domains-production.cloud.gov`.
+
+In other words, if you have a custom-domain service created for `directorate.agency.gov`, you'd want to create a CNAME record `_acme-challenge.directorate.agency.gov` with value `_acme-challenge.directorate.agency.gov.external-domains-production.cloud.gov`.
```
Comment (ccostino) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Clarifying language a bit:

```suggestion
For example, if you have a custom-domain service created for `directorate.agency.gov`, you'd create a CNAME record `_acme-challenge.directorate.agency.gov` with the value `_acme-challenge.directorate.agency.gov.external-domains-production.cloud.gov`.
```

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+`directorate.agency.cloud.gov.external-domains-production.cloud.gov`.
+
+Note that the second update here changes how users get to your site.
+We've made every effort to validate we're prepared for this change, but you should confirm that
+`$DOMAIN.external-domains-production.cloud.gov` currently resolves before making this change.
+
+You can do this by directly comparing the outputs of `nslookup $DOMAIN` and
+`nslookup $DOMAIN.external-domains-production.cloud.gov`, or by modifying your `hosts` file
+to point `$DOMAIN` to one of the IP addresses `$DOMAIN.external-domains-production.cloud.gov`
+currently resolves to. If either of these tests fail, STOP, do not update
```
Comment (ccostino) [inline _posts/2021-06-10-external-domain-migration-announcement.md]: Bolding an important point/step:

```suggestion
currently resolves to. **If either of these tests fail, _STOP_ and do not update**
```

**Review by ccostino — COMMENTED** [review]
Thank you, @bengerman13!  I had a few grammatical/language/formatting suggestions if they're helpful.

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+This post is for all users of the cdn-route or custom-domain services.
+
+## Background
+
+The custom-domain and cdn-route services leverage Let's Encrypt to provision certificates on our users' behalf.
+Several months ago, Let's Encrypt announced that they're deprecating their v1 API. To work with their new API, we've written a replacement
+to the cdn-route and custom-domain services, the external-domain service. We now need to migrate your service instances
+to the new service to ensure their certificates can continue renewing without issue. To accomplish this, we've written internal tooling to
+migrate instances automatically without service interruption, but it does require some action on your part to intiate the migration.
```
Comment (davemcorwin) [inline _posts/2021-06-10-external-domain-migration-announcement.md]:
```suggestion
migrate instances automatically without service interruption, but it does require some action on your part to initiate the migration.
```

**Review by davemcorwin — COMMENTED** [review]
Thanks @bengerman13 !!!

1. Do we need the trailing `.` in the domains?
2. Since these customers already have a `cdn-route` or `custom-domain` service, they will already have a DNS record for the user traffic, so the action they need to take is to edit or replace the current DNS record, not just create it.
3. Some folks may be using apex domains so may need an ALIAS record instead of a CNAME record.
4. is `nslookup` available on windows? Is there an alternative if not?
5. In the interest of plain language (I just took the training!) can we make this a little clearer using lists/tables? I think we can separate the details of the purpose of each DNS record from their values. Below is one take on it...

## What you need to do
When performing the following steps make sure to replace `$YOUR_DOMAIN` with your actual domain.

1. In your terminal or command line verify that `nsloookup $YOUR_DOMAIN` and `nslookup $YOUR_DOMAIN.external-domains-production.cloud.gov` return the same IP addresses. If they do not, _STOP_, do not update your DNS and contact cloud.gov support for assistance.

2. Create the following DNS record:

  | *type* | *name* | *value* |
  | --     | --     | --      |
  | CNAME  | _acme-challenge.$YOUR_DOMAIN | _acme-challenge.$YOUR_DOMAIN.external-domains-production.cloud.gov |

3. Edit the DNS record for $YOUR_DOMAIN to match:

| *type* | *name* | *value* |
| --     | --     | --      |
| CNAME  | $YOUR_DOMAIN | $YOUR_DOMAIN.external-domains-production.cloud.gov |

or, if your domain is a top-level domain:

| *type* | *name* | *value* |
| --     | --     | --      |
| ALIAS  | $YOUR_DOMAIN | $YOUR_DOMAIN.external-domains-production.cloud.gov |

4. Confirm you have the following DNS records

| *type* | *name* | *value* |
| --     | --     | --      |
| CNAME  | _acme-challenge.$YOUR_DOMAIN | _acme-challenge.$YOUR_DOMAIN.external-domains-production.cloud.gov |
| CNAME  | $YOUR_DOMAIN | $YOUR_DOMAIN.external-domains-production.cloud.gov |

or, if your domain is a top-level domain:

| *type* | *name* | *value* |
| --     | --     | --      |
| CNAME  | _acme-challenge.$YOUR_DOMAIN | _acme-challenge.$YOUR_DOMAIN.external-domains-production.cloud.gov |
| ALIAS  | $YOUR_DOMAIN | $YOUR_DOMAIN.external-domains-production.cloud.gov |

5. Confirm that your site is still available.

## More information
The CNAME record which begins with `_acme-challenge` allows us to provision SSL certificates on your behalf while the other CNAME or ALIAS record routes user traffic to your site.

-- fin

**Comment by bengerman13** [comment]
thanks @davemcorwin 
> Do we need the trailing . in the domains?

it technically exists on all DNS records. Whether you enter it or not depends on the tool you use to manage your DNS

> Since these customers already have a cdn-route or custom-domain service, they will already have a DNS record for the user traffic, so the action they need to take is to edit or replace the current DNS record, not just create it.

Right - I didn't spell this out because I thought it read better as creating 2 records vs create one and update one. Happy to change it if you think it's better the other way

> Some folks may be using apex domains so may need an ALIAS record instead of a CNAME record.

Also correct, probably worth spelling out. I'll update that

> is nslookup available on windows? Is there an alternative if not?

it is

> In the interest of plain language (I just took the training!) can we make this a little clearer using lists/tables? I think we can separate the details of the purpose of each DNS record from their values. Below is one take on it...

I think the table with both records is helpful, but the table in each step makes it hard to follow the list of steps - I'll incorporate the summary table.

**Review by onelittlebecca — DISMISSED** [review]
LGTM

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+The custom-domain and cdn-route services leverage Let's Encrypt to provision certificates on our users' behalf.
+Several months ago, Let's Encrypt announced that they're deprecating their v1 API. To work with their new API, we've written a replacement
+to the cdn-route and custom-domain services, the external-domain service. We now need to migrate your service instances
+to the new service to ensure their certificates can continue renewing without issue. To accomplish this, we've written internal tooling to
+migrate instances automatically without service interruption, but it does require some action on your part to initiate the migration.
+
+## What you need to do
+
+The new external-domain service uses a different method of validation with Let's Encrypt. The new method relies on specific DNS records being created for your domain(s). To begin the migration, you need to create 2 DNS CNAME records for each domain on your custom-domain or cdn-route instances.
```
Comment (davemcorwin) [inline _posts/2021-06-10-external-domain-migration-announcement.md]:
```suggestion
The new external-domain service uses a different method of validation with Let's Encrypt. The new method relies on specific DNS records being present for each of your domains. To begin the migration, you need to have configured the following DNS records for each domain on your custom-domain or cdn-route instances.
```

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+| Type  | Name                      | Value                                                           |
+|-------|---------------------------|-----------------------------------------------------------------|
+| CNAME | `_acme-challenge.$DOMAIN` | `_acme-challenge.$DOMAIN.external-domains-production.cloud.gov` |
+| CNAME | `$DOMAIN`                 | `$DOMAIN.external-domains-production.cloud.gov`                 |
```
Comment (davemcorwin) [inline _posts/2021-06-10-external-domain-migration-announcement.md]:
```suggestion
| CNAME* | `$DOMAIN`                 | `$DOMAIN.external-domains-production.cloud.gov`                 |
```

**Comment by onelittlebecca** [comment]
@bengerman13 Do we need to change this from draft PR  ?

File: _posts/2021-06-10-external-domain-migration-announcement.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+The `_acme-challenge.$DOMAIN` CNAME record allows us to provision SSL certificates on your behalf.
+
+As an example, if you have a custom-domain service created for `directorate.agency.gov`, you'd want to create a CNAME record `_acme-challenge.directorate.agency.gov` with value `_acme-challenge.directorate.agency.gov.external-domains-production.cloud.gov`.
+
+The `$DOMAIN` CNAME is responsible for routing the user traffic to your site.  Using the `directorate.agency.gov` example above, this would be `directorate.agency.gov.external-domains-production.cloud.gov`.
+
```
Comment (davemcorwin) [inline _posts/2021-06-10-external-domain-migration-announcement.md]:
```suggestion

If you already have `CNAME`, `A`, and/or `AAAA` (with `ALIAS`) record(s) for `$DOMAIN`, you should update the value or replace the record to match what is shown above.

* If your domain is a "2nd level domain" (i.e. `agency.gov` instead of `directorate.agency.gov`) you will need to use an `A` and/or `AAAA` (with `ALIAS`) record(s) assuming your DNS provider supports it.

```

**Comment by bengerman13** [comment]
I think we need to hold off until we've tested the migrator against some real instances

**Review by davemcorwin — COMMENTED** [review]
Added some specific suggestions to include references to A and AAAA records and try to clarify that existing records may need to be updated.

File: _posts/2021-06-10-external-domain-migration-announcement.md (hunk 10 above)
Comment (bengerman13) [reply _posts/2021-06-10-external-domain-migration-announcement.md]: I think this should be `and` rather than `and/or`

**Review by davemcorwin — APPROVED** [review]
Looks great to me, thanks @bengerman13 !!!!
